src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,100 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elements;
+    private int size;
+
     @Override
     public void add(T value) {
-
+        if (elements == null) {
+            elements = (T[]) new Object[DEFAULT_CAPACITY];
+        }
+        if (size == elements.length) {
+            elements = Arrays.copyOf(elements, elements.length + elements.length / 2);
+        }
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds");
+        }
+        if (elements == null) {
+            elements = (T[]) new Object[DEFAULT_CAPACITY];
+        }
+        if (size + 1 == elements.length) {
+            elements = Arrays.copyOf(elements, elements.length + elements.length / 2);
+        }
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds");
```

Comment:
repetative logic should be extracted to separate method (lines 48-49, 57-58, 65-67)